drivers/frequency/adf5355/adf5355.c

```diff
@@ -0,0 +1,518 @@
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include <inttypes.h>
+#include <stdio.h>
+#include "error.h"
+#include <malloc.h>
+#include "delay.h"
+#include "util.h"
+#include "adf5355.h"
+
+/******************************************************************************/
+/************************** Functions Implementation **************************/
+/******************************************************************************/
+
+/**
+ * SPI register write to device.
+ * @param dev - The device structure.
+ * @param reg - The register address.
+ * @param val - The register data.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+static int32_t adf5355_write(struct adf5355_dev *dev,
+			     uint8_t reg_addr,
+			     uint32_t data)
+{
+	uint8_t buf[ADF5355_SPI_NO_BYTES];
+	data = data | reg_addr;
+
+	buf[0] = data >> 24;
+	buf[1] = data >> 16;
+	buf[2] = data >> 8;
+	buf[3] = data;
+
+	return spi_write_and_read(dev->spi_desc, buf, ARRAY_SIZE(buf));
+}
+
+/**
+ * Compute PLL parameters.
+ * @param vco - The VCO frequency.
+ * @param pfd - The PFD frequency.
+ * @param integer - The integer division factor.
+ * @param fract1 - The fractionality.
+ * @param fract2 - The auxiliary fractionality.
+ * @param mod2 - The auxiliary modulus.
+ * @return None.
+ */
+static void adf5355_pll_fract_n_compute(uint64_t vco,
+					uint64_t pfd,
+					uint32_t *integer,
+					uint32_t *fract1,
+					uint32_t *fract2,
+					uint32_t *mod2,
+					uint32_t max_modulus2)
+{
+	uint64_t tmp;
+	uint32_t gcd_div;
+
+	tmp = do_div(&vco, pfd);
+	tmp = tmp * ADF5355_MODULUS1;
+	*fract2 = do_div(&tmp, pfd);
+
+	*integer = vco;
+	*fract1 = tmp;
+
+	*mod2 = pfd;
+
+	while (*mod2 > max_modulus2) {
+		*mod2 >>= 1;
+		*fract2 >>= 1;
+	}
+
+	gcd_div = greatest_common_divisor(*fract2, *mod2);
+	*mod2 /= gcd_div;
+	*fract2 /= gcd_div;
+}
+
+/**
+ * ADF5355 Register configuration
+ * @param dev - The device structure.
```

Comment:
nitpick: sync_all not documented. Not too bad, but might help with maintenance in the future.

Reply:
You are right. Will fix.